### PR TITLE
Fix query cache goleaks

### LINKED ISSUE #619: `TestNSMGR_RemoteUsecase_BusyEndpoints` has goroutine leaks sometimes
https://github.com/networkservicemesh/sdk/pull/612/checks?check_run_id=1508552180
```
--- FAIL: TestNSMGR_RemoteUsecase_BusyEndpoints (1.33s)
Error:     leaks.go:78: found unexpected goroutines:
        [Goroutine 638 in state chan send, with github.com/networkservicemesh/api/pkg/api/registry.ReadNetworkServiceChannel.func1 on top of the stack:
        goroutine 638 [chan send]:
        github.com/networkservicemesh/api/pkg/api/registry.ReadNetworkServiceChannel.func1(0x11b1760, 0xc0003ff440, 0xc00067e5a0)
        	/home/runner/go/pkg/mod/github.com/networkservicemesh/api@v0.0.0-20201204203731-4294f67deaa4/pkg/api/registry/helpers.go:54 +0x7a
        created by github.com/networkservicemesh/api/pkg/api/registry.ReadNetworkServiceChannel
        	/home/runner/go/pkg/mod/github.com/networkservicemesh/api@v0.0.0-20201204203731-4294f67deaa4/pkg/api/registry/helpers.go:52 +0x85
        ]
```

### PR DESCRIPTION
**Pull request by Bolodya1997**
# Issues
1. Query cache starts watching Find on chain context, it lasts until the application closes.
2. There is a possible leak when registry deletes NSE in process of cache Find.
# Solutions
1. Start watching Find on special context and close it after NSE remove.
2. Double check that NSE exists in registry before starting to wait events for it.

### CODE REVIEW

File: pkg/registry/common/querycache/nse_client.go [hunk 1]
```diff
@@ -82,9 +100,27 @@ func (q *queryCacheNSEClient) Find(ctx context.Context, in *registry.NetworkServ
...
+
+	client, err := next.NetworkServiceEndpointRegistryClient(ctx).Find(findCtx, nseQuery, opts...)
+	if err != nil {
+		return err
+	}
+
+	_, err = client.Recv()
+
+	return err
+}
```
Comment (denis-tingaikin) [inline pkg/registry/common/querycache/nse_client.go:122]: Could you explain why do we need this?

File: pkg/registry/common/querycache/nse_client.go [hunk 2]
```diff
@@ -42,34 +43,51 @@ func (q *queryCacheNSEClient) Find(ctx context.Context, in *registry.NetworkServ
...
 		go func() {
 			defer q.cache.Delete(key)
+
+			findCtx, findCancel := context.WithCancel(q.chainCtx)
+			defer findCancel()
+
 			nseQuery.Watch = true
-			stream, err := next.NetworkServiceEndpointRegistryClient(ctx).Find(q.chainCtx, nseQuery, opts...)
+
+			stream, err := next.NetworkServiceEndpointRegistryClient(ctx).Find(findCtx, nseQuery, opts...)
```
Comment (denis-tingaikin) [inline pkg/registry/common/querycache/nse_client.go:82]: findCtx is a nice catch :)

File: pkg/tools/sandbox/utils.go [hunk 3]
```diff
@@ -63,23 +66,30 @@ func NewEndpoint(ctx context.Context, nse *registry.NetworkServiceEndpoint, gene
...
 	for _, service := range nse.NetworkServiceNames {
 		if _, err := mgr.NetworkServiceRegistryServer().Register(ctx, &registry.NetworkService{Name: service, Payload: "IP"}); err != nil {
 			return nil, err
 		}
 	}
+
+	go func() {
+		<-ctx.Done()
+		for _, service := range nse.NetworkServiceNames {
+			_, _ = mgr.NetworkServiceRegistryServer().Unregister(context.Background(), &registry.NetworkService{Name: service, Payload: "IP"})
```
Comment (denis-tingaikin) [inline pkg/tools/sandbox/utils.go:86]: Let's replace "IP" with `payload.IP`

File: pkg/networkservice/chains/nsmgr/server_test.go [hunk 4]
```diff
@@ -308,28 +310,31 @@ func TestNSMGR_LocalUsecase(t *testing.T) {
 		},
 	}
 
-	conn, err := nsc.Request(ctx, request.Clone())
+	defer goleak.VerifyNone(t, goleak.IgnoreCurrent())
```
Comment (denis-tingaikin) [inline pkg/networkservice/chains/nsmgr/server_test.go:313]: Please use this at the top of func to prevent goleaks from the code above.

**Review by denis-tingaikin — CHANGES_REQUESTED** [review]

File: pkg/registry/common/querycache/nse_client.go (hunk 1 above)
Comment (Bolodya1997) [reply pkg/registry/common/querycache/nse_client.go]: I have removed `checkQuery`, but here is an issue: #678.

File: pkg/tools/sandbox/utils.go (hunk 3 above)
Comment (Bolodya1997) [reply pkg/tools/sandbox/utils.go]: Done.

File: pkg/networkservice/chains/nsmgr/server_test.go (hunk 4 above)
Comment (Bolodya1997) [reply pkg/networkservice/chains/nsmgr/server_test.go]: Oops, accidentally added this into the PR. Fixed.

File: pkg/tools/sandbox/utils.go [hunk 5]
```diff
@@ -60,29 +65,52 @@ func NewEndpoint(ctx context.Context, nse *registry.NetworkServiceEndpoint, gene
...
 		}
 	}
+
 	logger.Log(ctx).Infof("Started listen endpoint %v on %v.", nse.Name, u.String())
+
 	return &EndpointEntry{Endpoint: ep, URL: u}, nil
 }
 
+// UnregisterEndpoint unregisters endpoint from NSMgr
+func UnregisterEndpoint(ctx context.Context, nse *registry.NetworkServiceEndpoint, mgr nsmgr.Nsmgr) (err error) {
```
Comment (denis-tingaikin) [inline pkg/tools/sandbox/utils.go:99]: As I can see this func is using by one test. Can we make this incline?

Comment (Bolodya1997) [reply pkg/tools/sandbox/utils.go]: Yep, inlined.

**Review by denis-tingaikin — APPROVED** [review]

**Pull request merged**